pkg/util/taskgroup/taskgroup.go

```diff
@@ -0,0 +1,109 @@
+// Originally taken from https://github.com/ptxmac/multierrgroup
+
+// Package taskgroup provides a mix of multierr and errgroup
+// See documentation for https://pkg.go.dev/go.uber.org/multierr and https://pkg.go.dev/golang.org/x/sync/errgroup
+package taskgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// Group manages goroutines and collect all the errors.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#group for more information
+type Group interface {
+	WithContext(ctx context.Context) context.Context
+	WithPanicHandler(f func(any))
+	Wait() error
+	Go(name string, f func(logger *zap.Logger) error)
+}
+
+type group struct {
+	cancel       context.CancelFunc
+	logger       *zap.Logger
+	panicHandler func(any)
+
+	wg sync.WaitGroup
+
+	errMutex sync.Mutex
+	err      error
+}
+
+// NewGroup returns a new Group.
+func NewGroup(logger *zap.Logger) Group {
+	return &group{
+		cancel:       nil, // Set separately by WithContext
+		panicHandler: nil, // Set separately by WithPanicHandler
+		logger:       logger,
+		wg:           sync.WaitGroup{},
+
+		errMutex: sync.Mutex{},
+		err:      nil,
+	}
+}
+
+// WithContext updates the current Group with a associated Context.
+// The context will be canceled if any goroutine returns an error.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext
+func (g *group) WithContext(ctx context.Context) context.Context {
+	ctx, g.cancel = context.WithCancel(ctx)
+	return ctx
+}
+
+// WithPanicHandler sets a panic handler for the group.
+func (g *group) WithPanicHandler(f func(any)) {
+	g.panicHandler = f
+}
+
+// Wait blocks until all goroutines have completed.
+//
+// All errors returned from the goroutines will be combined into one using multierr and returned from this method.
+func (g *group) Wait() error {
+	g.wg.Wait()
+	if g.cancel != nil {
+		g.cancel()
+	}
+	return g.err
+}
+
+func (g *group) call(f func() error) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			if g.panicHandler != nil {
+				g.panicHandler(r)
+			}
+			err = fmt.Errorf("panic: %v", r)
```

Comment:
We should include a stack trace here. There's an implementation in `pkg/agent/runner.go` - if you like, `github.com/sharnoff/chord` also has some utils for this (but we also have no usage of it elsewhere, so... :shrug:)

Reply:
Oh, interesting, fair enough haha